features/assay_plate_verification.feature

```diff
@@ -64,9 +68,10 @@ Feature: Verify assay plate positions on the robot
   Scenario Outline: Barcodes scanned incorrectly on NX
     Given I have a process "Assay Plate creation" as part of the "Liquid handling" instrument with "Nx" assay plate verification
 
-    Given I can retrieve the plate with barcode "dest1" and parent barcodes "source1"
-    Given I can retrieve the plate with barcode "dest2" and parent barcodes "source1"
-    Given I cannot retrieve the plate with barcode "xxx"
+    Given I can retrieve plates by barcode
+      And I have a destination plate with barcode "dest1" and parent barcodes "source1"
+      And I have a destination plate with barcode "dest2" and parent barcodes "source1"
+      And I cannot retrieve the plate with barcode "xxx"
```

Comment:
Just to understand, why did these tests require a change in scenario structures?

Reply:
Good question.  The old code was querying plates one at a time from the V1 API when it wanted to find, day, 4 or 5 of them.  The new code queries them all in the same request which is better for the server, better performance and more robust because errors are less likely to occur mid process.  However, it does that by requesting them in an array and the mocking framework only lets you set up responses to the API for a specific set arguments.  Those mocks were generated upon each call of the plate with parent barcodes "Given"/"And" statement.  The actual code combines the request for all the destination plates and parent barcodes in one request, but you can't know when you're mocking that you need to combine them.

Hard to explain what I really mean here but as an example using the code you're commenting on:

* The first `And` statement says you want a `dest1` barcode and a `source1` parent barcode.  The mock is then set up to be able to query for the array `[ "source1", "dest1" ]`.  If that was the only instance of the destination plate `And` statement this would work because that's what the `ProcessPlate` model actually queries.
* The second `And` statement now also adds a `dest2` barcore and reuses the `source1` parent barcode.  The mock now gets set up to ALSO response to a query with the array `[ "source1", "dest2" ]`.
  * But the `ProcessPlate` model asks for the plates with `[ "source1", "dest1", "source1", "dest2" ]` which neither of the times you set up the mock would you know to mock for those arguments.

So instead, I added a `Given` that configures a mock for ANY arguments to get queried and it uses an in-memory list of registered plates to return the correct ones.  Now it wouldn't matter if you shuffled up that array of plates you want to find, it would behave like the real API.  The downside is that you do need to specify that you want to be able to retrieve plates as part of the test.  I tried to set up that mock in a `Before` block but the tests throw errors when you do that.  I think you can only set up the mocks during the actual scenario setup, and not before it.  I don't know and I felt like this was probably good enough and also indicates to the reader that you're activating a part of the mocked out Sequencescape API.

Reply:
Cool, Thanks much for the detailed explanation 👍